### PR TITLE
Improve testing framework for http registries

### PR DESCRIPTION
**Pull request by arlosi**
Improve integration of the http server introduced by the http-registry feature.

The same HTTP server is used for serving downloads, the index, and the API within the test framework.

This makes it easier to write tests that deal with authentication and http registries.

Most of this change is pulled from #10592. Getting it in separately should make the other change easier to review.

r? @Eh2406

### CODE REVIEW

**Comment by Eh2406** [comment]
I like it. @ehuss want to look it over before we +1?

File: src/cargo/core/source/source_id.rs [hunk 1]
```diff
@@ -680,7 +680,13 @@ impl<'a> fmt::Display for SourceIdAsUrl<'a> {
...
                 ref url,
                 ..
-            } => write!(f, "registry+{}", url),
+            } => {
+                if url.scheme().starts_with("sparse+") {
+                    write!(f, "{}", url)
+                } else {
+                    write!(f, "registry+{}", url)
+                }
+            }
```
Comment (epage) [inline src/cargo/core/source/source_id.rs:689]: Is this needed for the PR or an incidental artifact of being split out of another PR?

If its needed, should it be split out into a separate commit with a description of why this is needed?

Comment (arlosi) [reply src/cargo/core/source/source_id.rs:689]: It's not needed. I'll split it out into another PR and add a test.

File: tests/testsuite/publish.rs [hunk 2]
```diff
@@ -1522,14 +1529,20 @@ Caused by:
 ",
         )
         .run();
-
-    t.join().unwrap();
```
Comment (ehuss) [inline tests/testsuite/publish.rs:1526]: If I understand these changes correctly, the servers spawned by the tests are no longer shut down?  I'm concerned that if there are eventually dozens or hundreds of these that might cause issues. Is the fundamental problem not being able to interrupt the server waiting in `accept`?

File: tests/testsuite/credential_process.rs [hunk 3]
```diff
@@ -136,17 +140,14 @@ Only one of these values may be set, remove one or the other to proceed.
 ///
 /// Returns a thread handle for the API server, the test should join it when
 /// finished. Also returns the simple `foo` project to test against.
```
Comment (ehuss) [inline tests/testsuite/credential_process.rs:142]: Looks like the docs here might need to be updated.

File: tests/testsuite/publish.rs (hunk 2 above)
Comment (arlosi) [reply tests/testsuite/publish.rs:1526]: It attempts to shut the server down the `Drop for HttpServerHandle` (`registry.rs` L455), but we can't actually block on the thread joining because we're in `accept`

Comment (ehuss) [reply tests/testsuite/publish.rs:1526]: Would it be feasible to open a connection to the server to wake it up so that it can check `self.done` and exit?

Comment (arlosi) [reply tests/testsuite/publish.rs:1526]: That seems to work 👍

File: crates/cargo-test-support/src/registry.rs [hunk 4]
```diff
@@ -357,195 +419,323 @@ const DEFAULT_MODE: u32 = 0o644;
...
     fn drop(&mut self) {
-        self.done.store(true, Ordering::SeqCst);
-        // NOTE: we can't actually await the server since it's blocked in accept()
-        let _ = self.server.take();
+        if let Ok(mut stream) = TcpStream::connect(self.addr) {
+            // shutdown the server
+            stream.write_all(b"stop").unwrap();
+            stream.flush().unwrap();
+        }
+        let _ = self.join_handle.take().unwrap().join();
```
Comment (ehuss) [inline crates/cargo-test-support/src/registry.rs:459]: Can these unwraps be removed?  I'm concerned that if something goes wrong, it risks a panic-in-panic situation which causes an abort and can be difficult to diagnose.  For example, if a test fails, the panic exiting the test will drop the handle and then if one of these fails, it would cause an abort.

Also, is there a particular need to join with the server thread?  I would think it should be fine to assume the thread will exit (and to possibly avoid a deadlock if the server thread is somehow stuck).  I think the original code I wrote included `.join()` just to ensure that everything was OK. But in the context of a drop handler, I'd be a little uncomfortable with it since it will get called even on a failed test.  We're mostly concerned with `cargo` executing correctly, and not too concerned with the server.

Comment (arlosi) [reply crates/cargo-test-support/src/registry.rs:459]: The `join` was there to ensure the server was actually shutting down, but I can remove it.

**Comment by ehuss** [comment]
Thanks!

@bors r+
